src/gen_bpf.c

```diff
@@ -1346,6 +1487,83 @@ static struct bpf_blk *_gen_bpf_arch(struct bpf_state *state,
 		if (b_tail->next != NULL)
 			b_tail = b_tail->next;
 		blk_cnt++;
+		syscall_cnt++;
+
+		b_bintree = NULL;
+
+		/* build the binary tree if and else logic */
+		for (i = bintree_levels - 1; i >= 0; i--) {
+			level = SYSCALLS_PER_NODE << i;
+
+			if (((syscall_cnt + empty_cnt) % level) == 0) {
+				/* save the "if greater than" syscall num */
+				bintree_syscalls[i] = s_iter->num;
+				/* save the hash for the jf case */
+				bintree_hashes[i] = b_new->hash;
+
+				for (j = 0; j < i; j++) {
+					if (bintree_syscalls[j] == INVALID_SYSCALL ||
+					    bintree_hashes[j] == INVALID_HSH)
+						/* we are near the end of the binary
+						 * tree and the jump-to location is
+						 * not valid.  skip this if-else
+						 */
+						continue;
+
+					_BPF_INSTR(instr,
+						_BPF_OP(state->arch, BPF_JMP + BPF_JGT),
+						_BPF_JMP_NO,
+						_BPF_JMP_NO,
+						_BPF_K(state->arch, bintree_syscalls[j]));
+					instr.jt = _BPF_JMP_HSH(b_bintree == NULL ?
+							b_new->hash : b_bintree->hash);
+					instr.jf = _BPF_JMP_HSH(bintree_hashes[j]);
+					blk_cnt++;
+
+					b_bintree = _blk_append(state, NULL, &instr);
+					if (b_bintree == NULL)
+						goto arch_failure;
+					b_bintree->next = b_head;
+					if (b_head != NULL)
+						b_head->prev = b_bintree;
+					b_head = b_bintree;
+
+					rc = _hsh_add(state, &b_bintree, 1);
+					if (rc < 0)
+						goto arch_failure;
+				}
+
+				if (b_bintree != NULL)
+					/* this is the last if in this "block".
+					 * save it off so the next binary tree
+					 * if can "else" to it.
+					 */
+					bintree_hashes[j] = b_bintree->hash;
+				break;
+			}
```

Comment:
That entire "build the binary tree" for loop seems like a good candidate for moving into its own function, yes/no?

Reply:
Actually, looking at the code below, I'm not thinking that maybe the entire existing "create the syscall filters and add them to block list group" should be split out into separate optimization specific functions like you did with sort.

Thoughts?

Reply:
I agree - `_gen_bpf_arch` is way too monolithic now.

I'll see what I can do to divide it into some smaller more sane functions.

Reply:
I was able to split `_gen_bpf_arch()` up into a few smaller functions.  I think the division makes sense, but staring at that code too long may have warped my point of view ;)